### ISSUE TITLE
Unable to abort scan

### ISSUE THREAD

**Issue by ambarb** [NSLS-II/bluesky#348]
User ```^C^C``` after scan launch but before data collection began.  ```abort()``` has no effect.

````pytb
In [215]: ascan(theta,50,120,50)
Transient Scan ID: 58726
Persistent Unique Scan ID: '4c40a9b2-6f1f-408b-b467-66a75fdf09ef'
+------------+--------------+---------------------+----------------+-----------------------+
|   seq_num  |        time  |  fccd_stats5_total  |         theta  |  theta_user_setpoint  |
+------------+--------------+---------------------+----------------+-----------------------+
^C^C---------------------------------------------------------------------------
KeyboardInterrupt                         Traceback (most recent call last)
<ipython-input-215-d2feacb7c2bc> in <module>()
----> 1 ascan(theta,50,120,50)

/home/xf23id1/conda_envs/collection/lib/python3.4/site-packages/bluesky/simple_scans.py in __call__(self, motor, start, finish, intervals, time, subs, **kwargs)
    286         original_times = _set_acquire_time(time)
    287         result = super().__call__(motor, start, finish, intervals + 1,
--> 288                                   subs=subs, **kwargs)
    289         _unset_acquire_time(original_times)
    290         return result

/home/xf23id1/conda_envs/collection/lib/python3.4/site-packages/bluesky/simple_scans.py in __call__(self, subs, sub_factories, *args, **kwargs)
    182         # Any remainging kwargs go the RE. To be safe, no args are passed
    183         # to RE; RE args effectively become keyword-only arguments.
--> 184         return gs.RE(self.scan, _subs, **kwargs)
    185 
    186 

/home/xf23id1/conda_envs/collection/lib/python3.4/site-packages/bluesky/run_engine.py in __call__(self, plan, subs, **metadata_kw)
    558         with SignalHandler(signal.SIGINT) as self._sigint_handler:  # ^C
    559             self._task = loop.create_task(self._run())
--> 560             loop.run_forever()
    561             if self._task.done() and not self._task.cancelled():
    562                 exc = self._task.exception()

/home/xf23id1/conda_envs/collection/lib/python3.4/asyncio/base_events.py in run_forever(self)
    293         try:
    294             while True:
--> 295                 self._run_once()
    296                 if self._stopping:
    297                     break

/home/xf23id1/conda_envs/collection/lib/python3.4/asyncio/base_events.py in _run_once(self)
   1216                            timeout * 1e3, dt * 1e3)
   1217         else:
-> 1218             event_list = self._selector.select(timeout)
   1219         self._process_events(event_list)
   1220 

/home/xf23id1/conda_envs/collection/lib/python3.4/selectors.py in select(self, timeout)
    430             ready = []
    431             try:
--> 432                 fd_event_list = self._epoll.poll(timeout, max_ev)
    433             except InterruptedError:
    434                 return ready

KeyboardInterrupt: 

In [216]: abort()
Aborting....

In [217]: ascan(theta,70,120,50)
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-217-689ec3260b47> in <module>()
----> 1 ascan(theta,70,120,50)

/home/xf23id1/conda_envs/collection/lib/python3.4/site-packages/bluesky/simple_scans.py in __call__(self, motor, start, finish, intervals, time, subs, **kwargs)
    286         original_times = _set_acquire_time(time)
    287         result = super().__call__(motor, start, finish, intervals + 1,
--> 288                                   subs=subs, **kwargs)
    289         _unset_acquire_time(original_times)
    290         return result

/home/xf23id1/conda_envs/collection/lib/python3.4/site-packages/bluesky/simple_scans.py in __call__(self, subs, sub_factories, *args, **kwargs)
    182         # Any remainging kwargs go the RE. To be safe, no args are passed
    183         # to RE; RE args effectively become keyword-only arguments.
--> 184         return gs.RE(self.scan, _subs, **kwargs)
    185 
    186 

/home/xf23id1/conda_envs/collection/lib/python3.4/site-packages/bluesky/run_engine.py in __call__(self, plan, subs, **metadata_kw)
    534         # First thing's first: if we are in the wrong state, raise.
    535         if not self.state.is_idle:
--> 536             raise RuntimeError("The RunEngine is in a %s state" % self.state)
    537         if self._panic:
    538             raise PanicError("RunEngine is panicked. The run "

RuntimeError: The RunEngine is in a running state

In [218]: ascan(theta,70,120,50)
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-218-689ec3260b47> in <module>()
----> 1 ascan(theta,70,120,50)

/home/xf23id1/conda_envs/collection/lib/python3.4/site-packages/bluesky/simple_scans.py in __call__(self, motor, start, finish, intervals, time, subs, **kwargs)
    286         original_times = _set_acquire_time(time)
    287         result = super().__call__(motor, start, finish, intervals + 1,
--> 288                                   subs=subs, **kwargs)
    289         _unset_acquire_time(original_times)
    290         return result

/home/xf23id1/conda_envs/collection/lib/python3.4/site-packages/bluesky/simple_scans.py in __call__(self, subs, sub_factories, *args, **kwargs)
    182         # Any remainging kwargs go the RE. To be safe, no args are passed
    183         # to RE; RE args effectively become keyword-only arguments.
--> 184         return gs.RE(self.scan, _subs, **kwargs)
    185 
    186 

/home/xf23id1/conda_envs/collection/lib/python3.4/site-packages/bluesky/run_engine.py in __call__(self, plan, subs, **metadata_kw)
    534         # First thing's first: if we are in the wrong state, raise.
    535         if not self.state.is_idle:
--> 536             raise RuntimeError("The RunEngine is in a %s state" % self.state)
    537         if self._panic:
    538             raise PanicError("RunEngine is panicked. The run "

RuntimeError: The RunEngine is in a running state

In [219]: abort()
Aborting....

In [220]: abort()
Aborting....

In [221]: ascan(theta,70,120,50)
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-221-689ec3260b47> in <module>()
----> 1 ascan(theta,70,120,50)

/home/xf23id1/conda_envs/collection/lib/python3.4/site-packages/bluesky/simple_scans.py in __call__(self, motor, start, finish, intervals, time, subs, **kwargs)
    286         original_times = _set_acquire_time(time)
    287         result = super().__call__(motor, start, finish, intervals + 1,
--> 288                                   subs=subs, **kwargs)
    289         _unset_acquire_time(original_times)
    290         return result

/home/xf23id1/conda_envs/collection/lib/python3.4/site-packages/bluesky/simple_scans.py in __call__(self, subs, sub_factories, *args, **kwargs)
    182         # Any remainging kwargs go the RE. To be safe, no args are passed
    183         # to RE; RE args effectively become keyword-only arguments.
--> 184         return gs.RE(self.scan, _subs, **kwargs)
    185 
    186 

/home/xf23id1/conda_envs/collection/lib/python3.4/site-packages/bluesky/run_engine.py in __call__(self, plan, subs, **metadata_kw)
    534         # First thing's first: if we are in the wrong state, raise.
    535         if not self.state.is_idle:
--> 536             raise RuntimeError("The RunEngine is in a %s state" % self.state)
    537         if self._panic:
    538             raise PanicError("RunEngine is panicked. The run "

RuntimeError: The RunEngine is in a running state

In [222]: ascan(theta,70,120,50)
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-222-689ec3260b47> in <module>()
----> 1 ascan(theta,70,120,50)

/home/xf23id1/conda_envs/collection/lib/python3.4/site-packages/bluesky/simple_scans.py in __call__(self, motor, start, finish, intervals, time, subs, **kwargs)
    286         original_times = _set_acquire_time(time)
    287         result = super().__call__(motor, start, finish, intervals + 1,
--> 288                                   subs=subs, **kwargs)
    289         _unset_acquire_time(original_times)
    290         return result

/home/xf23id1/conda_envs/collection/lib/python3.4/site-packages/bluesky/simple_scans.py in __call__(self, subs, sub_factories, *args, **kwargs)
    182         # Any remainging kwargs go the RE. To be safe, no args are passed
    183         # to RE; RE args effectively become keyword-only arguments.
--> 184         return gs.RE(self.scan, _subs, **kwargs)
    185 
    186 

/home/xf23id1/conda_envs/collection/lib/python3.4/site-packages/bluesky/run_engine.py in __call__(self, plan, subs, **metadata_kw)
    534         # First thing's first: if we are in the wrong state, raise.
    535         if not self.state.is_idle:
--> 536             raise RuntimeError("The RunEngine is in a %s state" % self.state)
    537         if self._panic:
    538             raise PanicError("RunEngine is panicked. The run "

RuntimeError: The RunEngine is in a running state

In [223]: ascan?
Type:           AbsScan
String form:    <bluesky.simple_scans.AbsScan object at 0x7f873b492ba8>
File:           ~/conda_envs/collection/lib/python3.4/site-packages/bluesky/simple_scans.py
Signature:      ascan(self, motor, start, finish, intervals, time=None, subs=None, **kwargs)
Docstring:      ascan
Call signature: ascan(motor, start, finish, intervals, time=None, subs=None, **kwargs)
Call docstring:
Invoke the scan

Parameters
----------
motor
start : number
    The start point of the motion
finish : number
    The finish point of the motion
intervals : int
    The number of steps between `start` and `finish`
time : number
    The acquire time of the detector(s)?
subs : dict
    The temporary subscriptions to add to **this scan only**. These
    subscriptions are **not** persistent

In [224]: abort()
Aborting....

In [225]: abort()
Aborting....

In [226]: 
````

**Comment by danielballan** [comment]
That's a puzzler. I'll take another look at this logic. As usual any additional context you can provide helps -- anything that might have been different or unusual about this particular usage.

**Comment by ambarb** [comment]
User was scanning with fccd with 4 seconds per image. He did many other previous scans exactly as the one aborted. Data collection had not yet started.

**Comment by tacaswell** [comment]
It seems the problem is that (for what ever reason) the Ctrtl-C escaped our clutches, which left the run engine in a bad state.  I think we need to do two things:

 - [ ] have abort be a bit more aggressive about cleaning up for un-expected states
 - [ ] throw is some more try..except block in

From the trace back it looks like the exception was raised in the ayncio select logic, not in code we wrote which means there is some sort of race-condition / bad time thing going on :worried: .

**Comment by danielballan** [comment]
This was, we think, fixed by #379. Hard to confirm, but I think we can close this.
